File: backend/src/main/java/com/campuspass/backend/model/Advertisement.java
package com.campuspass.backend.model;

import com.campuspass.backend.model.enums.AdPosition;
import com.campuspass.backend.model.enums.AdStatus;
import jakarta.persistence.*;
import java.time.LocalDate;
import java.time.LocalDateTime;

@Entity
@Table(name = "advertisements")
public class Advertisement {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private Long merchantId;

    private String title;
    private String description;
    /** Texte du bouton (ex. bandeau accueil HOME_TOP). */
    @Column(name = "cta_label")
    private String ctaLabel;
    private String imageUrl;
    private String videoUrl;
    private String targetUrl;

    /**
     * Ciblage simple : ville (ex: Ouagadougou).
     */
    private String targetCity;

    /** Ciblage simple : pays (ex: Burkina Faso). */
    private String targetCountry;

    /**
     * Ciblage simple : université (nom ou code).
     */
    private String targetUniversity;

    /**
     * Segment utilisateur ciblé (ALL, SUBSCRIBED, NON_SUBSCRIBED, etc.).
     * Interprété côté service/mobile.
     */
    private String targetSegment;

    @Enumerated(EnumType.STRING)
    private AdPosition position;

    private LocalDate startDate;
    private LocalDate endDate;
    /** Budget campagne (FCFA). */
    private Double budget;
    /** ID de l'offre liée (pour SPONSORED_OFFER). */
    private Long offerId;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private AdStatus status = AdStatus.ACTIVE;

    private LocalDateTime createdAt;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "merchantId", insertable = false, updatable = false)
    private Merchant merchant;

    public Advertisement() {}

    public Long getId() { return id; }
    public void setId(Long id) { this.id = id; }
    public Long getMerchantId() { return merchantId; }
    public void setMerchantId(Long merchantId) { this.merchantId = merchantId; }
    public String getTitle() { return title; }
    public void setTitle(String title) { this.title = title; }
    public String getDescription() { return description; }
    public void setDescription(String description) { this.description = description; }
    public String getCtaLabel() { return ctaLabel; }
    public void setCtaLabel(String ctaLabel) { this.ctaLabel = ctaLabel; }
    public String getImageUrl() { return imageUrl; }
    public void setImageUrl(String imageUrl) { this.imageUrl = imageUrl; }

    public String getVideoUrl() { return videoUrl; }
    public void setVideoUrl(String videoUrl) { this.videoUrl = videoUrl; }

    public String getTargetUrl() { return targetUrl; }
    public void setTargetUrl(String targetUrl) { this.targetUrl = targetUrl; }
    public String getTargetCity() { return targetCity; }
    public void setTargetCity(String targetCity) { this.targetCity = targetCity; }
    public String getTargetCountry() { return targetCountry; }
    public void setTargetCountry(String targetCountry) { this.targetCountry = targetCountry; }
    public String getTargetUniversity() { return targetUniversity; }
    public void setTargetUniversity(String targetUniversity) { this.targetUniversity = targetUniversity; }
    public String getTargetSegment() { return targetSegment; }
    public void setTargetSegment(String targetSegment) { this.targetSegment = targetSegment; }
    public AdPosition getPosition() { return position; }
    public void setPosition(AdPosition position) { this.position = position; }
    public LocalDate getStartDate() { return startDate; }
    public void setStartDate(LocalDate startDate) { this.startDate = startDate; }
    public LocalDate getEndDate() { return endDate; }
    public void setEndDate(LocalDate endDate) { this.endDate = endDate; }
    public AdStatus getStatus() { return status; }
    public void setStatus(AdStatus status) { this.status = status; }
    public LocalDateTime getCreatedAt() { return createdAt; }
    public void setCreatedAt(LocalDateTime createdAt) { this.createdAt = createdAt; }
    public Double getBudget() { return budget; }
    public void setBudget(Double budget) { this.budget = budget; }
    public Long getOfferId() { return offerId; }
    public void setOfferId(Long offerId) { this.offerId = offerId; }
}

File: campuspass_app/lib/models/advertisement.dart
class Advertisement {
  final int id;
  final int merchantId;
  final String title;
  final String? imageUrl;
  final String? videoUrl;
  final String? description;
  /// Texte du bouton (bandeau accueil HOME_TOP, etc.).
  final String? ctaLabel;
  final String position; // HOME_BANNER, HOME_TOP, SPONSORED_OFFER, OFFERS_PAGE, SEARCH_PAGE, NOTIFICATION
  final int? offerId;
  final String? targetUrl;
  final String? startDate;
  final String? endDate;
  final String status; // ACTIVE, INACTIVE...
  final double? budget;
  final String? createdAt;

  Advertisement({
    required this.id,
    required this.merchantId,
    required this.title,
    this.imageUrl,
    this.videoUrl,
    this.description,
    this.ctaLabel,
    required this.position,
    this.offerId,
    this.targetUrl,
    this.startDate,
    this.endDate,
    required this.status,
    this.budget,
    this.createdAt,
  });

  factory Advertisement.fromJson(Map<String, dynamic> json) {
    // Backend parfois en snake_case: on rend le parsing tolérant.
    String? asString(dynamic v) => v == null ? null : v.toString();
    int asInt(dynamic v) => (v is num) ? v.toInt() : int.parse(v.toString());

    final imageUrl = asString(json['imageUrl'] ?? json['image_url'] ?? json['image']);
    final videoUrl = asString(json['videoUrl'] ?? json['video_url'] ?? json['video']);
    final targetUrl = asString(json['targetUrl'] ?? json['target_url'] ?? json['link']);
    final title = asString(json['title'] ?? json['name'] ?? json['label']) ?? '';
    final description = asString(json['description'] ?? json['text'] ?? json['subtitle']);
    final ctaLabel = asString(json['ctaLabel'] ?? json['cta_label']);
    final offerId = json['offerId'] ?? json['offer_id'];

    return Advertisement(
      id: asInt(json['id']),
      merchantId: asInt(json['merchantId'] ?? json['merchant_id']),
      title: title,
      imageUrl: imageUrl,
      videoUrl: videoUrl,
      description: description,
      ctaLabel: ctaLabel,
      position: json['position'] as String? ?? 'HOME_TOP',
      offerId: offerId == null ? null : asInt(offerId),
      targetUrl: targetUrl,
      startDate: json['startDate']?.toString(),
      endDate: json['endDate']?.toString(),
      status: json['status'] as String? ?? 'ACTIVE',
      budget: (json['budget'] as num?)?.toDouble(),
      createdAt: json['createdAt']?.toString(),
    );
  }
}

